src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20240802123234_Routine_RebuildIndexes.sql

```diff
@@ -0,0 +1,283 @@
+    [ReportSent] DATETIME2,
+  );
+
+  IF OBJECT_ID(N'__RebuildIndexesAlterIndexes', N'U') IS NULL
+  CREATE TABLE __RebuildIndexesAlterIndexes
+  (
+    [Id]                 INT              NOT NULL,
+    [RunId]              UNIQUEIDENTIFIER NOT NULL,
+    [IndexName]          NVARCHAR(MAX)    NOT NULL,
+    [SchemaName]         NVARCHAR(MAX)    NOT NULL,
+    [ObjectName]         NVARCHAR(MAX)    NOT NULL,
+    [StartTime]          DATETIME2,
+    [EndTime]            DATETIME2,
+    [InitialFragPercent] INT,
+    [StartFragPercent]   INT,
+    [EndFragPercent]     INT,
+    [Info]               NVARCHAR(MAX),
+  );
+
+  IF OBJECT_ID(N'__RebuildIndexesUpdateStatistics', N'U') IS NULL
+  CREATE TABLE __RebuildIndexesUpdateStatistics
+  (
+    [RunId]     UNIQUEIDENTIFIER NOT NULL,
+    [TableName] NVARCHAR(MAX)    NOT NULL,
+    [StartTime] DATETIME2,
+    [EndTime]   DATETIME2,
+  );
+
+  -- Create new row for this stored procedure run
+  INSERT INTO __RebuildIndexes (Id, StartTime, EndTime, HitTimeout, ReportSent)
+  VALUES (@RunId, GETUTCDATE(), NULL, NULL, NULL);
+
+  -- Create rows for indexes
+  SELECT @StatsCount = COUNT(Id) FROM #Stats;
+
+  SET @StatsRowIndex = 1;
+
+  WHILE @StatsRowIndex <= @StatsCount
+    BEGIN
+      INSERT INTO __RebuildIndexesAlterIndexes
+      SELECT @StatsRowIndex,
+             @RunId,
+             IndexName,
+             SchemaName,
+             ObjectName,
+             NULL,
+             NULL,
+             FragPercent,
+             NULL,
+             NULL,
+             NULL
+      FROM #Stats
+      WHERE Id = @StatsRowIndex;
+
+      SET @StatsRowIndex += 1;
+    END
+
+  -- Create rows for tables
+  INSERT INTO __RebuildIndexesUpdateStatistics
+  SELECT @RunId,
+         TL.ObjectName,
+         NULL,
+         NULL
+  FROM #TableList TL;
+
+  /*-----------------------------------------------------------------------*/
+
+  -- Rebuild or reorganize each index depending on the fragmentation percentage
+  SET @StatsRowIndex = 1;
+
+  WHILE @StatsRowIndex <= @StatsCount AND @SkipRemainingTasks = 0
+    BEGIN
+      DECLARE @FragPercentBefore INT;
+      SELECT @Command =
+             'ALTER INDEX ' + IndexName + ' ON ' + SchemaName + '.' + ObjectName +
+             IIF(FragPercent >= @FragmentationThresholdRebuild,
+                 ' REBUILD WITH (ONLINE = ON)',
+                 ' REORGANIZE'),
+             @FragPercentBefore = FragPercent
+      FROM #Stats
+      WHERE Id = @StatsRowIndex;
+
+      IF @FragPercentBefore < @FragmentationThresholdReorganize
+        BEGIN
+          RAISERROR ('Skipping command - fragmentation percent did not hit reorganisation threshold ''%s''.', 0, 1, @Command) WITH NOWAIT;
+
+          UPDATE __RebuildIndexesAlterIndexes
+          SET Info = 'Skipped: fragmentation percentage did not hit reorganisation threshold'
+          WHERE RunId = @RunId
+            AND Id = @StatsRowIndex;
+
+          SET @StatsRowIndex += 1;
+          CONTINUE;
+        END
```

Comment:
I'm inclined to leave it - mostly to save me the time rejigging it. 

My preference is towards guard clauses because it's already confusing what's happening with all the indenting - although I do agree it's easy to get caught out not incrementing `@StatsRowIndex` - in fact I think I had that bug when developing this.

Reply:
Cool, yeah I only left this comment because it requires more careful managing of the counter.